### PR TITLE
Fix std.range.retro not behaving the same as foreach_reverse

### PR DESCRIPTION
**Pull request by iK4tsu**
This was first discussed on the Discord server, and it was only about the 'save' method being enforced. This PR touches only on that, but since then I've been thinking and the whole behavior should be the same as the 'foreach_reverse'.

@maxhaton said at the time he was going to analyze the code, so I would like to know his opinion as well.

### CODE REVIEW

File: std/range/package.d
Comment (ljmf00) [inline std/range/package.d:243]: There is no need for multiple import statements of the same module.

**Review by ljmf00 — CHANGES_REQUESTED** [review]
Otherwise LGTM

File: std/range/package.d
Comment (iK4tsu) [reply std/range/package.d:243]: The way it was before was harder to read IMO.

**Comment by maxhaton** [comment]
This seems good to my eyes but I'd like some approval from Atila or Andrei.

**Comment by iK4tsu** [comment]
I would like to discuss this further also. IMO I think retro should account for `popBack`, `back`, and `empty` as constraints, and have the internal implementation rely on what range actually is.

**Comment by iK4tsu** [comment]
Tagging @atilaneves @andralex

File: std/range/package.d
Comment (RazvanN7) [reply std/range/package.d:243]: This change does not follow the usual guidelines in phobos. I suggest undoing it.

**Comment by iK4tsu** [comment]
> I think that this change is making matters more confusing. Right now we have a clear definition of range types and the algorithms in phobos are based on that definition. With this PR, we special case retro for Frankenranges that are InputRanges but also have the back and popBack properties. I am against such an addition because this will open the door for complicating other phobos functions to support Frankenranges.

I see what you mean, but in these very specific cases shouldn't we loosen a bit? We have a retro range that the only thing it does is iterating back to front, but it can only iterate on ranges that are Bidirectional compliant. If it is properly documented that the `retro` range would follow the `foreach_reverse` semantics I don't think it would be an issue.

**Comment by ljmf00** [comment]
> I think that this change is making matters more confusing. Right now we have a clear definition of range types and the algorithms in phobos are based on that definition. With this PR, we special case `retro` for Frankenranges that are InputRanges but also have the `back` and `popBack` properties.

I kinda see your point here, but a few things to note here. First, this Frankenrange should be the definition of `isBidirectionalRange`. AFAIK, there is no rationale to add a `save` method to a bidirectional range other than the hierarchy implemented by @andralex in the ranges interfaces.

> I am against such an addition because this will open the door for complicating other phobos functions to support Frankenranges.

These cases where this is wrong is pretty much because `isBidirectionalRange` is poorly designed. There is no reason to enforce the user to implement a `save` method when 90% of the time is left unused.

**Comment by andralex** [comment]
>I see what you mean, but in these very specific cases shouldn't we loosen a bit?

I'd opine not. This is the wrong kind of generality and I argued in the forum several times, one that I remember is the [Generality creep](https://forum.dlang.org/post/q6plhj$1l9$1@digitalmars.com) thread.

This code is exactly one step on the proverbial road to hell paved with good intentions.

`retro` works for bidirectional ranges, period.

**Comment by andralex** [comment]
> AFAIK, there is no rationale to add a save method to a bidirectional range other than the hierarchy implemented by @andralex in the ranges interfaces.

[Abstraction is selective ignorance](https://jonjagger.blogspot.com/2010/08/models-and-abstraction.html). Therefore, there will be corner cases and situations not covered properly by an abstraction. This is to be expected and understood as a cost of doing business when defining abstractions.

Forward ranges represent ranges that have state stored somewhere (most often in memory, but can be conceivably in files, too). This is fundamentally what allows one to save the state of iteration and use different ranges to iterate the same data. The data is not produced on the fly, even if it may seem to (as is the case for random number generators - they do have stored state without representing the entire range in memory).

Now bidirectional ranges allow iteration from both ends, which is the moral equivalent of allowing two ranges over the same underlying data. That means again that the range must have state stored somewhere, so it must be a forward range, too.

I'm not saying it's impossible to conceive a range that is bidirectional yet cannot be multiply iterated. It's just that that case, if it ever exists, is narrow and uninteresting enough to be ignored for the sake of simpler abstractions.

**Comment by ljmf00** [comment]
> `retro` works for bidirectional ranges, period.

It should, but the bidirectional range definition is just wrong.

> [Abstraction is selective ignorance](https://jonjagger.blogspot.com/2010/08/models-and-abstraction.html). Therefore, there will be corner cases and situations not covered properly by an abstraction. This is to be expected and understood as a cost of doing business when defining abstractions.

I can't really understand your point with this other than a bidirectional range being a bloated abstraction. It is bloated because functions like `retro` enforce the user to have a `save` even if the user doesn't want to. For example, in order to use `retro` with a struct that doesn't need additional logic for saving state, the user needs to this pointless method:

```d
R save() => this;
```

Even if that struct has additional logic to save state, `retro` only need to optionally export it, to be reiterated.

There is a clear difference between what you referenced here. This abstraction is not strictly needed and for corner cases, these methods can be implemented, only when needed.

> Forward ranges represent ranges that have state stored somewhere (most often in memory, but can be conceivably in files, too). This is fundamentally what allows one to save the state of iteration and use different ranges to iterate the same data. The data is not produced on the fly, even if it may seem to (as is the case for random number generators - they do have stored state without representing the entire range in memory).

I understand the point of forward ranges, it makes sense, even if the naming is a bit weird, but enforcing a bidirectional range to be forward range is just non-sense.

> Now bidirectional ranges allow iteration from both ends, which is the moral equivalent of allowing two ranges over the same underlying data. That means again that the range must have state stored somewhere, so it must be a forward range, too.
> 
> I'm not saying it's impossible to conceive a range that is bidirectional yet cannot be multiply iterated. It's just that that case, if it ever exists, is narrow and uninteresting enough to be ignored for the sake of simpler abstractions.

`retro` and possibly other functions do the same thing because of this design. `retro` literally does nothing with `save` other than exporting it in the result range. It is bad design enforcing someone to implement a method that will possibly never ever be used.

---

I can see this PR as a hacky workaround to the bidirectional range definition but at least makes `retro` a more functional function to use, in terms of its practicality, and solving the root of this problem would be a tremendous breaking change.

**Comment by iK4tsu** [comment]
> Abstraction is selective ignorance. Therefore, there will be corner cases and situations not covered properly by an abstraction. This is to be expected and understood as a cost of doing business when defining abstractions.

Ok, I understand. But does that justify that every design choice must follow the abstraction? Shouldn't the abstraction serve as a guiding line to help create a more standerdized final product, but not completely take out of the equation approaches that make more sense in these said corner cases?

> which is the moral equivalent of allowing two ranges over the same underlying data.

Can you elaborate more on this statement? I got a bit lost here, I may do not understand completely what you mean by this.

**Comment by andralex** [comment]
> > `retro` works for bidirectional ranges, period.
> 
> It should, but the bidirectional range definition is just wrong.

That's a bit much, but I agree experience has brought about several improvements we could bring to the range interface. That would need to be deferred to a future, incompatible version of the standard library.

> > [Abstraction is selective ignorance](https://jonjagger.blogspot.com/2010/08/models-and-abstraction.html). Therefore, there will be corner cases and situations not covered properly by an abstraction. This is to be expected and understood as a cost of doing business when defining abstractions.
> 
> I can't really understand your point with this other than a bidirectional range being a bloated abstraction. It is bloated because functions like `retro` enforce the user to have a `save` even if the user doesn't want to.

By the same token, users are required to define `popFront` even if "they don't want to" because their coding standards prescribe `pop_front` or `PopFront`.

I agree `save` is a trivial function, but that's irrespective to `retro`. We introduced `save` to make it easy to distinguish input ranges from forward ranges by API, a luxury that C++ input iterators and forward iterators don't have and are paying dearly for. If I redid the API now, I'd probably just prescribe forward ranges to add some `bool` constant `isForwardRange` and nothing else. But even then, there will be the occasional user who "doesn't want to" add that constant.

> I understand the point of forward ranges, it makes sense, even if the naming is a bit weird, but enforcing a bidirectional range to be forward range is just non-sense.

That's again a bit much. "Combativeness should be in proportion to righteousness" and all that.  Having bidirectional ranges be a superset of forward ranges allows using a bunch of algorithms designed for forward ranges transparently with bidirectional ranges, with no extra hoops to jump. Designing a library that works with a mishmash of almost-clear-but-not-quite-clear relationships between categories is unnecessarily difficult.

**Comment by andralex** [comment]
> > Abstraction is selective ignorance. Therefore, there will be corner cases and situations not covered properly by an abstraction. This is to be expected and understood as a cost of doing business when defining abstractions.
> 
> Ok, I understand. But does that justify that every design choice must follow the abstraction?

Only that design choices that break the abstractions come with exceptional costs and should be approached with trepidation.

> Shouldn't the abstraction serve as a guiding line to help create a more standardized final product, but not completely take out of the equation approaches that make more sense in these said corner cases?

Um, no.

> > which is the moral equivalent of allowing two ranges over the same underlying data.
> 
> Can you elaborate more on this statement? I got a bit lost here, I do not understand completely what you mean by this.

One can go through the same data with front/popFront as well as back/popBack. That means almost always that the data is supported persistently so it's appropriate to use with forward iteration, too.

**Comment by iK4tsu** [comment]
> By the same token, users are required to define popFront even if "they don't want to" because their coding standards prescribe pop_front or PopFront.

That is different. Those methods are defined in the language spec (https://dlang.org/spec/statement.html#foreach-with-ranges), the range hierarchy is a Phobos implementation.

**Comment by iK4tsu** [comment]
> Having bidirectional ranges be a superset of forward ranges allows using a bunch of algorithms designed for forward ranges transparently with bidirectional ranges, with no extra hoops to jump.

I feel like the whole hierarchy system is too restrictive. Wouldn't it be better if ranges separated modularly and each range implementation would have as constraints **only** the pieces that were really needed and then implementing the support for the optional parts depending on what range it was?

I feel that could work, but most likely it was already thought and led to this:

> Designing a library that works with a mishmash of almost-clear-but-not-quite-clear relationships between categories is unnecessarily difficult.

But even if range implementations are to follow only the Phobos range hierarchy implementation, I still think we should have a module that supports this and other similar cases. The user shouldn't be forced to implement methods just to be able to work with the API when it doesn't make sense to have them in their use case. Or implementing their own Ranges because they have no support in the standard library, even though that implementation makes sense. In this case, it is not too hard to implement it, but it becomes an inconvenience to do it. I feel this gives a bit of friction to the use of Ranges since the language behaviors allow to use them in a certain way that the Phobos implementations do not.

**Comment by ljmf00** [comment]
> By the same token, users are required to define `popFront` even if "they don't want to" because their coding standards prescribe `pop_front` or `PopFront`.

No that's different. `popFront` is a standard interface, like, `front`, `popBack` and even `save`. The naming doesn't interfere with actually enforcing the usage of it.

> I agree `save` is a trivial function, but that's irrespective to `retro`. We introduced `save` to make it easy to distinguish input ranges from forward ranges by API, a luxury that C++ input iterators and forward iterators don't have and are paying dearly for.

Introducing C++ STL here is not a good way to go forward. STL is criticized by many people and not a good thing for comparison to take, IMO.

> If I redid the API now, I'd probably just prescribe forward ranges to add some `bool` constant `isForwardRange` and nothing else. But even then, there will be the occasional user who "doesn't want to" add that constant.

I think that is worse, but I also think we are shifting a bit of what the PR is proposing. Either way, I already expressed my opinion on this. If we look at other sane languages, this is not being enforced for a simple thing like `reverse` or `retro` function. The implementation of a `foreach_reverse` is a whole tricky workaround to circumvent this and D end up not providing a functional-ready way to do it.

I understand the point @RazvanN7 raised about "Frankenranges" but I think we also need to look at other aspects and the actual root of the problem. Doing a dummy `save` to use `retro` is just... weird to say the least.

**Comment by andralex** [comment]
> > By the same token, users are required to define popFront even if "they don't want to" because their coding standards prescribe pop_front or PopFront.
> 
> That is different. Those methods are defined in the language spec (https://dlang.org/spec/statement.html#foreach-with-ranges), the range hierarchy is a Phobos implementation.

Point taken, thanks.

**Comment by andralex** [comment]
> > I agree `save` is a trivial function, but that's irrespective to `retro`. We introduced `save` to make it easy to distinguish input ranges from forward ranges by API, a luxury that C++ input iterators and forward iterators don't have and are paying dearly for.
> 
> Introducing C++ STL here is not a good way to go forward. STL is criticized by many people and not a good thing for comparison to take, IMO.

We're not "introducing STL here"; it has been a major influence of the ranges design from the outset. If you have a dim view of the STL you probably wouldn't enjoy Phobos either.

> > If I redid the API now, I'd probably just prescribe forward ranges to add some `bool` constant `isForwardRange` and nothing else. But even then, there will be the occasional user who "doesn't want to" add that constant.
> 
> I think that is worse,

This makes me curious. It would be cool if you shared what would be better. 

> but I also think we are shifting a bit of what the PR is proposing. Either way, I already expressed my opinion on this. If we look at other sane languages, this is not being enforced for a simple thing like `reverse` or `retro` function.

What languages are those, and what aspects of them or their standard libraries are pertinent to this discussion?

At any rate, looking at the current state of affairs and finding it entirely unsatisfactory is the engine of progress. You seem to be confident to have a grasp on how to do ranges (or a similar abstraction) radically better than the STL or Phobos. That would be very interesting - I encourage you to put together such an API and open it for discussion on the forums.

**Comment by andralex** [comment]
> > Having bidirectional ranges be a superset of forward ranges allows using a bunch of algorithms designed for forward ranges transparently with bidirectional ranges, with no extra hoops to jump.
> 
> I feel like the whole hierarchy system is too restrictive. Wouldn't it be better if ranges separated modularly and each range implementation would have as constraints **only** the pieces that were really needed and then implementing the support for the optional parts depending on what range it was?

Interesting. I discussed this a bit in a DConf talk and IIRC elsewhere. One crushing problem with C++ concepts (which are somewhat competing with D's constraints) is that every concepts needs a name, leading to an explosion of awkward names. In D you can actually express concepts on the fly by mentioning expressions that need to work inside the constraint.

Ranges already do something of that. We have input ranges, forward ranges etc. but we also have `hasLength` which is orthogonal to the range categories. So at least conceviable we could define an algorithm that takes input ranges that have length.

Related: Even today we have a number of significant ranges that have a trivial `empty` that always returns `false`. Then why should we ever define it, as opposed to simply assuming a range not defining it is never empty.

The golden ratio is probably between these two extremes. Having algorithms specify all primitives they need every time would be exhausting, so some simple abstraction names ("I need a bidirectional range") are welcome. The other extreme would have us define names such as `BidirectionalRangeWithSaveAndLength` which is clearly bad.

All in all what's now in place is workable. A redesign would probably be more so because we'd derive from experience.

> I feel that could work, but most likely it was already thought and led to this:
> 
> > Designing a library that works with a mishmash of almost-clear-but-not-quite-clear relationships between categories is unnecessarily difficult.
> 
> But even if range implementations are to follow only the Phobos range hierarchy implementation, I still think we should have a module that supports this and other similar cases. The user shouldn't be forced to implement methods just to be able to work with the API when it doesn't make sense to have them in their use case. Or implementing their own Ranges because they have no support in the standard library, even though that implementation makes sense. In this case, it is not too hard to implement it, but it becomes an inconvenience to do it. I feel this gives a bit of friction to the use of Ranges since the language behaviors allow to use them in a certain way that the Phobos implementations do not.

This seems a fabricated problem. I don't think someone who goes to the trouble of creating a bidirectional range with all the primitives but then refuses to define `save` is much deserving of support.

**Comment by iK4tsu** [comment]
> One crushing problem with C++ concepts (which are somewhat competing with D's constraints) is that every concept needs a name, leading to an explosion of awkward names. In D you can actually express concepts on the fly by mentioning expressions that need to work inside the constraint.

Naming wise only gets weird when trying to word Range combinations. If left to the individual aspect I don't see that being an issue. There would be base Ranges and then you would combine them for each implementation. I want a RangeA that has B and C aspects. Trying to name that Range combination might be awkward, but using it separately isn't and is explicit.

> Related: Even today we have a number of significant ranges that have a trivial empty that always returns false. Then why should we ever define it, as opposed to simply assuming a range not defining it is never empty.

I might be related, but once again it won't be a valid argument. This is defined in the language spec.

I will add to this that I like the idea of Rust's ranges incorporated with `Option`. IIRC they don't need an `empty` for iteration because it is checked by the output directly `Some` or `None`, they just need to keep iterating until a `None` is returned. This also prevents `out of bounds` access since `front` is always accessible.

> This seems a fabricated problem. I don't think someone who goes to the trouble of creating a bidirectional range with all the primitives but then refuses to define save is much deserving of support.

It's now fabricated. I never said refusing. What I meant was, even though this can be a solution it feels like a workaround for something that isn't permitted by Phobos. Even though this can be fixed, I think in conversation we've agreed on this isn't an ideal solution and should be addressed in the future.

I agree that changing the current behavior of `std.range` that relies heavily on Phobos's Range hierarchy would be bitter-sweet and could lead to some unwanted arguments for future PRs addressing something similar. I also think there should be support for this in Phobos that's why I argued, as a second solution, the implementation of an alternative function in another module.

**Comment by RazvanN7** [comment]
Is it ok to close this?

**Comment by ljmf00** [comment]
> We're not "introducing STL here"; it has been a major influence of the ranges design from the outset. If you have a dim view of the STL you probably wouldn't enjoy Phobos either.

I guess that's is one of the problems. Yes I have a dim view of STL and a lot of people does have really good points against some STL API interfaces. I like the overall design of ranges but it may need some improvements and that is probably taken from the experience of using it, as you mentioned before.

> This makes me curious. It would be cool if you shared what would be better.

I would rather create a DoubleEnded iterator or maintain the name of bidirectional range and get rid of `save` by optionally requiring it. If you want a bidirectional range that are also forward range, you just combine those two.

> What languages are those, and what aspects of them or their standard libraries are pertinent to this discussion?

I won't extensively show examples to prove my point but looking at languages like Python and Rust can be a good thing to be based on. The usage of `next`, for example, instead of `popFront`, `front` and `empty` could be a thing to improve ranges API, at least, providing a generic `next` in Phobos using those three functions could be beneficial for functional programming.

The Rust `rev` implementation uses a `DoubleEndedIterator` which is in practice a bidirectional range without `save` by default. You can look at Rust documentation [here](https://doc.rust-lang.org/std/iter/trait.DoubleEndedIterator.html) and [here](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.rev). This is one of the things I would make Phobos a better standard library additionally, in combination with `Result` and `Optional` types, but I heard that someone is against `Optional` type, for some reason, not sure if it is actually you and, whoever is, why.

Similarly, Python implements a `reversed` function requiring the described protocols in the function documentation. See more [here](https://docs.python.org/3/library/functions.html#reversed).

> Related: Even today we have a number of significant ranges that have a trivial `empty` that always returns `false`. Then why should we ever define it, as opposed to simply assuming a range not defining it is never empty.

It is to explicitly tell that the range is infinite. If that was specified in the infinite range definition, that could also be optionally defined.

> The golden ratio is probably between these two extremes. Having algorithms specify all primitives they need every time would be exhausting, so some simple abstraction names ("I need a bidirectional range") are welcome. The other extreme would have us define names such as `BidirectionalRangeWithSaveAndLength` which is clearly bad.

Why? By definition, being bidirectional doesn't imply that. You just added that to the D definition of bidirectional range.

> All in all what's now in place is workable. A redesign would probably be more so because we'd derive from experience.

The point of this PR would be to take a step forward for that change.

> This seems a fabricated problem. I don't think someone who goes to the trouble of creating a bidirectional range with all the primitives but then refuses to define `save` is much deserving of support.

I'm rather appealing to make things simpler, just that.

**Comment by RazvanN7** [comment]
Improving the range hierarchy is a topic that should be thoroughly discussed via a DIP. Such PRs that modify a small piece of phobos to accommodate for some use cases that might make sense at a first glance should be treated cautiously. Therefore, I am going to close this PR, but if you will pursue this endeavor, feel free to reopen once you have obtained the community's approval.

**Comment by ljmf00** [comment]
> Improving the range hierarchy is a topic that should be thoroughly discussed via a DIP. Such PRs that modify a small piece of phobos to accommodate for some use cases that might make sense at a first glance should be treated cautiously. Therefore, I am going to close this PR, but if you will pursue this endeavor, feel free to reopen once you have obtained the community's approval.

The points presented about range hierarchy are crucial to make Phobos a better standard library and the current behaviour of some functions such as `retro` are unnecessarily bloated. This PR has a solution for this and permits the users to use `retro` in a more sane and simple way, otherwise, people need to make their own implementation of `retro` or add a useless `save` function because they are obligated to.

Anyway, I think a DIP is hard to accept due be a breaking change, although if @iK4tsu wants to push this forward I would really appreciate the effort on this.

**Pull request closed without merging**
